### PR TITLE
[None][feat] Enable CUDA graph support for KvConnectorWorker API

### PR DESCRIPTION
**Pull request by nv-kmcgill53**
## Summary by CodeRabbit

* **New Features**
  * Added support for registering a forward-pass callback that runs immediately after computation, enabling timely cache management/offloading.
  * Enabled CUDA Graphs for scenarios using the KV cache connector, removing prior limitations and allowing graph-captured execution with callbacks.
  * Forward-pass notifications now work consistently across both eager and CUDA Graphs paths, improving throughput and reducing stalls during cache offloading.

## Description

This PR enables NVIDIA Dynamo's KV Block Manager to register a forward pass cuda event and associated callback through the connector worker api. 

Upon finishing the forward pass, TRTLLM now optionally records a user supplied cuda event on the same stream as the forward pass. Immediately thereafter, the user's provided callback is invoked. which (optionally) waits on the event.

In conjunction with: https://github.com/ai-dynamo/dynamo/pull/3297

## PR Checklist

Please review the following before submitting your PR:
- PR description clearly explains what and why. If using CodeRabbit's summary, please make sure it makes sense.
- PR Follows [TRT-LLM CODING GUIDELINES](https://github.com/NVIDIA/TensorRT-LLM/blob/main/CODING_GUIDELINES.md) to the best of your knowledge.
- Test cases are provided for new code paths (see [test instructions](https://github.com/NVIDIA/TensorRT-LLM/tree/main/tests#1-how-does-the-ci-work))
- Any new dependencies have been scanned for license and vulnerabilities
- [CODEOWNERS](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/CODEOWNERS) updated if ownership changes
- Documentation updated as needed
- The reviewers assigned automatically/manually are appropriate for the PR.

- [x] Please check this after reviewing the above items as appropriate for this PR.

## GitHub Bot Help

`/bot [-h] ['run', 'kill', 'skip', 'reuse-pipeline'] ...`

Provide a user friendly way for developers to interact with a Jenkins server.

Run `/bot [-h|--help]` to print this help message.

See details below for each supported subcommand.

<details>

`run  [--reuse-test (optional)pipeline-id --disable-fail-fast --skip-test --stage-list "A10-PyTorch-1, xxx" --gpu-type "A30, H100_PCIe" --test-backend "pytorch, cpp" --add-multi-gpu-test --only-multi-gpu-test --disable-multi-gpu-test --post-merge --extra-stage "H100_PCIe-TensorRT-Post-Merge-1, xxx" --detailed-log --debug(experimental)]`

Launch build/test pipelines. All previously running jobs will be killed.

`--reuse-test (optional)pipeline-id ` *(OPTIONAL)* : Allow the new pipeline to reuse build artifacts and skip successful test stages from a specified pipeline or the last pipeline if no pipeline-id is indicated. If the Git commit ID has changed, this option will be always ignored. The DEFAULT behavior of the bot is to reuse build artifacts and successful test results from the last pipeline.

`--disable-reuse-test ` *(OPTIONAL)* : Explicitly prevent the pipeline from reusing build artifacts and skipping successful test stages from a previous pipeline. Ensure that all builds and tests are run regardless of previous successes.

`--disable-fail-fast ` *(OPTIONAL)* : Disable fail fast on build/tests/infra failures.

`--skip-test ` *(OPTIONAL)* : Skip all test stages, but still run build stages, package stages and sanity check stages. Note: Does **NOT** update GitHub check status.

`--stage-list "A10-PyTorch-1, xxx"` *(OPTIONAL)* : Only run the specified test stages. Examples: "A10-PyTorch-1, xxx". Note: Does **NOT** update GitHub check status.

`--gpu-type "A30, H100_PCIe"` *(OPTIONAL)* : Only run the test stages on the specified GPU types. Examples: "A30, H100_PCIe". Note: Does **NOT** update GitHub check status.

`--test-backend "pytorch, cpp"` *(OPTIONAL)* : Skip test stages which don't match the specified backends. Only support [pytorch, cpp, tensorrt, triton]. Examples: "pytorch, cpp" (does not run test stages with tensorrt or triton backend). Note: Does **NOT** update GitHub pipeline status.

`--only-multi-gpu-test ` *(OPTIONAL)* : Only run the multi-GPU tests. Note: Does **NOT** update GitHub check status.

`--disable-multi-gpu-test ` *(OPTIONAL)* : Disable the multi-GPU tests. Note: Does **NOT** update GitHub check status.

`--add-multi-gpu-test ` *(OPTIONAL)* : Force run the multi-GPU tests in addition to running L0 pre-merge pipeline.

`--post-merge ` *(OPTIONAL)* : Run the L0 post-merge pipeline instead of the ordinary L0 pre-merge pipeline.

`--extra-stage "H100_PCIe-TensorRT-Post-Merge-1, xxx"` *(OPTIONAL)* : Run the ordinary L0 pre-merge pipeline and specified test stages. Examples: --extra-stage "H100_PCIe-TensorRT-Post-Merge-1, xxx".

`--detailed-log ` *(OPTIONAL)* : Enable flushing out all logs to the Jenkins console. This will significantly increase the log volume and may slow down the job.

`--debug ` *(OPTIONAL)* : **Experimental feature**. Enable access to the CI container for debugging purpose. Note: Specify exactly one stage in the `stage-list` parameter to access the appropriate container environment. Note: Does **NOT** update GitHub check status.

For guidance on mapping tests to stage names, see `docs/source/reference/ci-overview.md`
and the `scripts/test_to_stage_mapping.py` helper.

### kill

`kill  `

Kill all running builds associated with pull request.

### skip

`skip --comment COMMENT `

Skip testing for latest commit on pull request. `--comment "Reason for skipping build/test"` is required. IMPORTANT NOTE: This is dangerous since lack of user care and validation can cause top of tree to break.

### reuse-pipeline

`reuse-pipeline `

Reuse a previous pipeline to validate current commit. This action will also kill all currently running builds associated with the pull request. IMPORTANT NOTE: This is dangerous since lack of user care and validation can cause top of tree to break.

</details>

### CODE REVIEW

**Review by jthomson04 — CHANGES_REQUESTED** [review]

**Review by jthomson04 — APPROVED** [review]

File: tensorrt_llm/_torch/pyexecutor/model_engine.py [hunk 1]
```diff
@@ -2309,6 +2314,10 @@ def capture_forward_fn(inputs: Dict[str, Any]):
                         outputs = self.cuda_graph_runner.replay(
                             batch_size, inputs)
 
+            if self.forward_pass_callback is not None:
+                self.forward_pass_callback.event.record()
+                self.forward_pass_callback.callback()
```
[the remark this answers is not in the document]
Comment (nv-kmcgill53) [reply tensorrt_llm/_torch/pyexecutor/model_engine.py]: @pcastonguay Here is the print out of the current stream which `replay()` is called on and the stream which we are recording and synchronizing the event on. They are all the same:
```
  1849 Current stream inside replay: <function current_stream at 0x76e14c77afc0>
   1850 [10/15/2025-00:32:52] [TRT-LLM] [I] Current CUDA stream is: <function current_stream at 0x76e14c77afc0>
   1851 stream which callable is recording event on is: <function current_stream at 0x76e14c77afc0>
```
Since there are no direct APIs in torch which allow me to query this and put in asserts, I think this is the best we can do? Let me know if have a more robust way of testing or a corner case which I need to think about.

File: tensorrt_llm/_torch/pyexecutor/kv_cache_connector.py [hunk 2]
```diff
@@ -81,6 +81,17 @@ class SchedulerOutput:
...
+    class ForwardPassCallback:
+
+        def __init__(self, event: torch.cuda.Event):
+            self.event = event
+
+        def callback(self):
+            """
+            Define the work needed to be done when the forward_pass_event
+            is triggered.
+            """
```
[the remark this answers is not in the document]
Comment (nv-kmcgill53) [reply tensorrt_llm/_torch/pyexecutor/kv_cache_connector.py]: updated the api to record the event within the callback

File: tensorrt_llm/_torch/pyexecutor/model_engine.py [hunk 3]
```diff
@@ -469,6 +471,10 @@ def __init__(
         else:
             self.cache_indirection_attention = None
 
+    def register_forward_pass_callback(
+            self, callback: KvCacheConnectorWorker.ForwardPassCallback):
+        self.forward_pass_callback = callback
```
[the remark this answers is not in the document]
Comment (nv-kmcgill53) [reply tensorrt_llm/_torch/pyexecutor/model_engine.py]: Updated the API to be more generic to expect, and hold on to, a callable which will be called at the end of the forward pass. KVBM will record the event and wait on the event ourselves. The way this is implemented, we have the same overhead as we do today, we synchronize on the event in the same stream as the forward pass. We can look at a better long term solution after this change gets checked in.

**Review by pcastonguay — APPROVED** [review]

File: tensorrt_llm/_torch/pyexecutor/kv_cache_connector.py [hunk 4]
```diff
@@ -95,6 +95,18 @@ def get_connector_meta(self) -> object:
     def _clear_connector_meta(self):
         self._metadata = None
 
+    def register_forward_pass_callable(self) -> Callable:
```
Comment (nv-kmcgill53) [inline tensorrt_llm/_torch/pyexecutor/kv_cache_connector.py]: @pcastonguay Just a heads up, I removed the `@abstractmethod` tag on this function to make the class backward compatible with earlier implementations of this class. I think KVBM is the only benefactor of this, but it also benefits other customers who may be making use of this.

**Review by thorjohnsen — APPROVED** [review]
LGTM.
